pkg/etcdutil/etcdutil.go

```diff
@@ -294,3 +321,87 @@ func TakeAndSaveFullSnapshot(ctx context.Context, client client.MaintenanceClose
 
 	return snapshot, nil
 }
+
+// checkFullSnapshotIntegrity verifies the integrity of the full snapshot by comparing
+// the appended SHA256 hash of the full snapshot with the calculated SHA256 hash of the full snapshot data.
+func checkFullSnapshotIntegrity(snapshotData io.ReadCloser, snapTempDBFilePath string, logger *logrus.Entry) (io.ReadCloser, error) {
+	logger.Info("checking the full snapshot integrity with the help of SHA256")
+
+	// If previous temp db file already exist then remove it.
+	if err := os.Remove(snapTempDBFilePath); err != nil && !os.IsNotExist(err) {
+		return nil, err
+	}
+
+	db, err := os.OpenFile(snapTempDBFilePath, os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err := io.Copy(db, snapshotData); err != nil {
+		return nil, err
+	}
+
+	lastOffset, err := db.Seek(0, io.SeekEnd)
+	if err != nil {
+		return nil, err
+	}
+	// 512 is chosen because it's a minimum disk sector size in most systems.
+	hasHash := (lastOffset % 512) == sha256.Size
+	if !hasHash {
+		return nil, fmt.Errorf("SHA256 hash seems to be missing from snapshot data")
+	}
+
+	totalSnapshotBytes, err := db.Seek(-sha256.Size, io.SeekEnd)
+	if err != nil {
+		return nil, err
+	}
+
+	// get snapshot SHA256 hash
+	sha := make([]byte, sha256.Size)
+	if _, err := db.Read(sha); err != nil {
+		return nil, fmt.Errorf("failed to read SHA256 from snapshot data %v", err)
+	}
+
+	buf := make([]byte, hashBufferSize)
+	hash := sha256.New()
+
+	logger.Infof("Total no. of bytes received from snapshot api call with SHA: %d", lastOffset)
+	logger.Infof("Total no. of bytes received from snapshot api call without SHA: %d", totalSnapshotBytes)
+
+	// reset the file pointer back to starting
+	currentOffset, err := db.Seek(0, io.SeekStart)
+	if err != nil {
+		return nil, err
+	}
+
+	for currentOffset+hashBufferSize <= totalSnapshotBytes {
+		offset, err := db.Read(buf)
+		if err != nil {
+			return nil, fmt.Errorf("unable to read snapshot data into buffer to calculate SHA256: %v", err)
+		}
+
+		hash.Write(buf[:offset])
+		currentOffset += int64(offset)
+	}
+
+	if currentOffset < totalSnapshotBytes {
+		if _, err := db.Read(buf); err != nil {
+			return nil, fmt.Errorf("unable to read last chunk of snapshot data into buffer to calculate SHA256: %v", err)
+		}
+
+		hash.Write(buf[:totalSnapshotBytes-currentOffset])
+	}
+
+	dbSha := hash.Sum(nil)
+	if !bytes.Equal(sha, dbSha) {
+		return nil, fmt.Errorf("expected SHA256 for full snapshot: %v, got %v", sha, dbSha)
+	}
+
+	// reset the file pointer back to starting
+	if _, err := db.Seek(0, io.SeekStart); err != nil {
+		return nil, err
+	}
+
+	// full-snapshot of database has been successfully verified.
+	return db, nil
+}
```

Comment:
> // resetFilePointer resets the file pointer to the beginning of the file.
func resetFilePointer(db *os.File) error {
	if _, err := db.Seek(0, io.SeekStart); err != nil {
		return fmt.Errorf("failed to reset file pointer: %w", err)
	}
	return nil
}
// cleanUpTempFile removes the temporary file used during snapshot verification.
func cleanUpTempFile(filePath string, logger *logrus.Entry) {
	if err := os.Remove(filePath); err != nil && !os.IsNotExist(err) {
		logger.Warnf("Failed to remove temporary full snapshot file: %v", err)
	}
}

I wouldn't like to create a separate function for one liner.

Reply:
> 	buf := make([]byte, hashBufferSize)
	if _, err := io.CopyBuffer(db, snapshotData, buf); err != nil {
		return nil, fmt.Errorf("failed to copy snapshot data to temporary DB file %s: %v", snapTempDBFilePath, err)
	}
	// Verify SHA256 hash
	if err := validateSnapshotSHA256(db, logger); err != nil {
		return nil, err
	}
	// Reset the file pointer back to the beginning
	if err := resetFilePointer(db); err != nil {
		return nil, err
	}


this is wrong code, this is not how it's working.

Reply:
May I know what you mean by wrong.

Reply:
as we have to read the data of snapshots in buffer and keep calculating the hash

```go
	buf := make([]byte, hashBufferSize)
	hash := sha256.New()

	// reset the file pointer back to starting
	currentOffset, err := db.Seek(0, io.SeekStart)
	if err != nil {
		return nil, err
	}

	for currentOffset+hashBufferSize <= totalSnapshotBytes {
		offset, err := db.Read(buf)
		if err != nil {
			return nil, fmt.Errorf("unable to read snapshot data into buffer to calculate SHA256: %v", err)
		}

		hash.Write(buf[:offset])
		currentOffset += int64(offset)
	}
```

I see you are using `io.LimitReader` but IMO we want to avoid that as we don't want to load all db data (~8Gi in worst case) into memory, that's why we went with this approach of reading the data in small chunk, so that there won't be any memory spike as you have yourself has verified this: https://github.com/gardener/etcd-backup-restore/pull/779#issuecomment-2425647408

Reply:
`io.LimitReader` does not load all the data (~8Gi in the worst case) into memory at once. Instead, it ensures that at most `totalSnapshotBytes` are read from the underlying `db`, and the actual memory usage is determined by the buffer size provided to `io.CopyBuffer` (in this case, `hashBufferSize`). This keeps the memory footprint small and controlled while simplifying the code by removing manual offset tracking and loop management.

If there are still concerns about the behavior or specific cases where `io.LimitReader` might not meet expectations, I’d be happy to investigate further. Let me know if you’d like me to profile this or make adjustments!